libraries/Networking/src/NetworkInterface.cpp

```diff
@@ -0,0 +1,719 @@
+        return IPAddress();
+    }
+    return IPAddress(d.ip.u_addr.ip4.addr);
+}
+
+IPAddress NetworkInterface::broadcastIP() const
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return IPAddress();
+    }
+    return calculateBroadcast(IPAddress(ip.gw.addr), IPAddress(ip.netmask.addr));
+}
+
+IPAddress NetworkInterface::networkID() const
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return IPAddress();
+    }
+    return calculateNetworkID(IPAddress(ip.gw.addr), IPAddress(ip.netmask.addr));
+}
+
+uint8_t NetworkInterface::subnetCIDR() const
+{
+    if(_esp_netif == NULL){
+        return (uint8_t)0;
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return (uint8_t)0;
+    }
+    return calculateSubnetCIDR(IPAddress(ip.netmask.addr));
+}
+
+IPAddress NetworkInterface::linkLocalIPv6() const
+{
+    if(_esp_netif == NULL){
+        return IPAddress(IPv6);
+    }
+    static esp_ip6_addr_t addr;
+    if(esp_netif_get_ip6_linklocal(_esp_netif, &addr)){
+        return IPAddress(IPv6);
+    }
+    return IPAddress(IPv6, (const uint8_t *)addr.addr, addr.zone);
+}
+
+IPAddress NetworkInterface::globalIPv6() const
+{
+    if(_esp_netif == NULL){
+        return IPAddress(IPv6);
+    }
+    static esp_ip6_addr_t addr;
+    if(esp_netif_get_ip6_global(_esp_netif, &addr)){
+        return IPAddress(IPv6);
+    }
+    return IPAddress(IPv6, (const uint8_t *)addr.addr, addr.zone);
+}
+
+size_t NetworkInterface::printTo(Print & out) const {
+    size_t bytes = 0;
+    bytes += out.print(esp_netif_get_desc(_esp_netif));
+    bytes += out.print(":");
+    if(esp_netif_is_netif_up(_esp_netif)){
+        bytes += out.print(" <UP");
+    } else {
+        bytes += out.print(" <DOWN");
+    }
+    bytes += printDriverInfo(out);
+    bytes += out.println(">");
+
+    bytes += out.print("      ");
+    bytes += out.print("ether ");
+    bytes += out.print(macAddress());
+    bytes += out.println();
+
+    bytes += out.print("      ");
+    bytes += out.print("inet ");
+    bytes += out.print(localIP());
+    bytes += out.print(" netmask ");
+    bytes += out.print(subnetMask());
+    bytes += out.print(" broadcast ");
+    bytes += out.print(broadcastIP());
+    bytes += out.println();
+
+    bytes += out.print("      ");
+    bytes += out.print("gateway ");
+    bytes += out.print(gatewayIP());
+    bytes += out.print(" dns ");
+    bytes += out.print(dnsIP());
+    bytes += out.println();
+
+    static const char * types[] = { "UNKNOWN", "GLOBAL", "LINK_LOCAL", "SITE_LOCAL", "UNIQUE_LOCAL", "IPV4_MAPPED_IPV6" };
+    esp_ip6_addr_t if_ip6[CONFIG_LWIP_IPV6_NUM_ADDRESSES];
+    int v6addrs = esp_netif_get_all_ip6(_esp_netif, if_ip6);
```

Comment:
It would be nice to have a way to access these if needed, e.g. `allIPv6(uint8_t ipv6_no)`, very similar to `dnsIP()`.

The function just calls esp_netif_get_all_ipv6() and if ipv6_no > v6addrs (or zero) then returns the empty default address. Not very efficient if you loop through them all, but better than nothing.

(Not sure if ipv6_no should be 0-based (normal) or 1-based (like `dnsIP()`).  Maybe call it `ipv6_index` for a 0-based array, i.e. 0 returns `IPAddress(IPv6, (const uint8_t *)if_ip6[0].addr, if_ip6[0].zone)` and so on.

Reply:
you do have access. All you need to get them is the esp_netif, which you get through `ETH.netif()`, `WiFi.STA.netif()` or `WiFi.AP.netif()`. Just not "Arduino" level API, as such is too early to be decided.

Reply:
That's fine then.

Reply:
I like the work so far; it is making good progress.